File: cpp/leetcode/260.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& nums) {
        int res = 0;
        for(auto num : nums) res ^= num;
        int maker = 1;
        while(!(res & maker)) {
            maker <<= 1;
        }
        int x = 0;
        for(auto num : nums){
            if(num & maker) x ^= num;
        }
        return {x, res ^ x};
    }
};

File: python/leetcode/260.py
class Solution(object):
    def singleNumber(self, nums):
        res = 0
        for item in nums: res ^= item
        # res is x ^ y
        # find the digit where x != y
        maker = 1
        while (maker & res) != maker:
            maker <<= 1
        x = 0
        for item in nums:
            # divide the num into 2 group, one has the same digit as x
            if maker & item:  x ^= item   # get the x value
        return [x, x ^ res]